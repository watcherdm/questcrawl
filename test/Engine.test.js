require('./mock.js')
const { QuestCrawlCard } = require('../src/c_Card.js')
const { getPartyToken } = require('../src/c_Party')
const { Grid } = require('../src/c_Grid.js')
const { TurnOrder, recallCards } = require('../src/c_TurnOrder.js')

const {
  onPartyMoved,
  init,
  generateIsland,
  createObj,
  resetBoard,
  toScreenGrid,
  resetRng
} = require('../src/b_Engine.js')
describe('Engine', () => {
  beforeAll(() => {
    createObj('character', { name: 'Party', avatar: 'test.url', _id: 'valid-character-id' })
    createObj('graphic', {
      name: 'Party',
      imgsrc: 'test.url',
      _id: 'valid-graphic-id-1',
      pageid: global.getPageId(),
      layer: 'objects',
      height: 70,
      width: 70
    })
  })

  it('should provide the init method', () => {
    expect(typeof init).toBe('function')
  })
  describe('init', () => {
    it('should set the global grid and turnorder', () => {
      const state = {
        QuestCrawl: {
          grid: false,
          turnorder: false
        }
      }
      init(state)
      expect(global.grid).toBeInstanceOf(Grid)
      expect(global.turnorder).toBeInstanceOf(TurnOrder)
    })
  })

  describe('generateIsland', () => {
    describe('home town already exists', () => {
      let mockShuffleDeck
      beforeEach(() => {
        const state = {
          QuestCrawl: {
            grid: [{ x: 0, y: 0, id: 'hometowngraphic', cardid: 'red-joker-1' }],
            turnorder: false
          }
        }
        mockShuffleDeck = jest.spyOn(global, 'shuffleDeck')
        init(state)
        expect(global.grid).toBeInstanceOf(Grid)
        expect(global.turnorder).toBeInstanceOf(TurnOrder)
      })
      afterEach(() => {
        jest.restoreAllMocks()
      })
      it('should exit if the grid already has a home town', () => {
        generateIsland('invalid-deck-id', [])
        expect(mockShuffleDeck).not.toHaveBeenCalled()
      })
    })
    describe('empty grid', () => {
      beforeEach(() => {
        createObj('card', { id: 'hometown-id', deckid: 'valid-deckid', name: 'Red Joker' })
        const state = {
          QuestCrawl: {
            grid: [],
            turnorder: false
          }
        }
        init(state)
        global.grid.setup = true
        global.placed = []
        recallCards()
      })
      it('should not throw', () => {
        expect(() => generateIsland('valid-deckid', [])).not.toThrow()
      })
      it('should draw the hometown card', () => {
        const drawCardSpy = jest.spyOn(global, 'drawCard')
        generateIsland('valid-deckid', [])
        expect(drawCardSpy).toHaveBeenCalled()
      })
      it('should call place on the hometown card', () => {
        const placeSpy = jest.spyOn(QuestCrawlCard.prototype, 'place')
        generateIsland('valid-deckid', [])
        expect(placeSpy).toHaveBeenCalledWith(true)
      })
      it('should set 54 placed cards', () => {
        generateIsland('valid-deckid', [])
        expect(global.placed.length).toBe(54)
      })
    })
    describe('known seed', () => {
      beforeEach(() => {
        const state = {
          QuestCrawl: {
            grid: [],
            turnorder: false
          }
        }
        init(state)
        resetRng('test-seed')
        global.grid.setup = true
        global.placed = []
        recallCards()
      })
      describe('static', () => {
        it('should return the same grid when static is used', () => {
          generateIsland('valid-deckid', ['--static 1'])
          expect(global.grid.toJSON()).toMatchSnapshot()
        })
      })
      describe('not static', () => {
        it('should place different cards in the same locations', async () => {
          generateIsland('valid-deckid', [])
          const grid1 = global.grid.toJSON()
          await resetBoard()
          resetRng('test-seed')
          generateIsland('valid-deckid', [])
          const grid2 = global.grid.toJSON()
          expect(grid1.map(({ x, y }) => `${x}:${y}`).sort()).toStrictEqual(grid2.map(({ x, y }) => `${x}:${y}`).sort())
          expect(grid1).not.toStrictEqual(grid2)
        })
      })
    })
  })

  describe('onPartyMoved', () => {
    it('should exist as a function', () => {
      expect(typeof onPartyMoved).toBe('function')
    })
    describe('when invoked with party token', () => {
      beforeEach(() => {
        createObj('card', { id: 'hometown-id', deckid: 'valid-deckid', name: 'Red Joker' })
        global.state = {
          QuestCrawl: {
            config: {
              mode: 'original'
            },
            grid: [],
            turnorder: false,
            history: [{ x: 0, y: 0 }]
          }
        }
        return resetBoard().then(() => {
          init(state)
          global.grid.setup = true
          global.placed = []
          recallCards()
          generateIsland('valid-deckid', ['--gaps 0'])
        }).catch((e) => log(e))
      })

      it('should not throw an error', () => {
        expect(() => onPartyMoved(global.grid, getPartyToken())).not.toThrow()
      })
      it('should not allow the party to move to a disconnected territory', () => {
        const partyToken = getPartyToken()
        onPartyMoved(global.grid, partyToken)
        expect(partyToken.get('left')).toBe(1085)
        expect(partyToken.get('top')).toBe(1085)
      })
      it('should allow the party to move to a connected territory', () => {
        const partyToken = getPartyToken()
        const hometown = global.grid.get({ x: 0, y: 0 })
        let moveto = hometown.getRandomNeighbor()
        while (moveto.id.indexOf('-') !== 0) {
          moveto = hometown.getRandomNeighbor()
        }
        onPartyMoved(global.grid, partyToken)
        const newCoords = toScreenGrid(moveto, 70)
        partyToken.set({
          ...newCoords
        })
        onPartyMoved(global.grid, partyToken)
        expect(partyToken.get('left')).toBe(newCoords.left)
      })
      it('should chat the card contents', () => {
        const chatCardSpy = jest.spyOn(global, 'chatCard')
        const partyToken = getPartyToken()
        const hometown = global.grid.get({ x: 0, y: 0 })
        let moveto = hometown.getRandomNeighbor()
        while (moveto.id.indexOf('-') !== 0) {
          moveto = hometown.getRandomNeighbor()
        }
        onPartyMoved(global.grid, partyToken)
        partyToken.set({
          ...toScreenGrid(moveto, 70)
        })
        onPartyMoved(global.grid, partyToken)
        expect(chatCardSpy).toHaveBeenCalled()
      })
    })
  })
})
