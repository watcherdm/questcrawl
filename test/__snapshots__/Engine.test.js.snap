// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Engine generateIsland known seed static should return the same grid when static is used 1`] = `
[
  {
    "blank": false,
    "cardid": "-card-test-id-52",
    "face": "red",
    "faceup": true,
    "id": "-graphic-test-id-52",
    "name": "Red Joker",
    "suit": "joker",
    "x": 0,
    "y": 0,
  },
  {
    "cardid": "Lake",
    "faceup": true,
    "id": "Gap",
    "x": 1,
    "y": 0,
  },
  {
    "blank": false,
    "cardid": "-card-test-id-5",
    "face": "two",
    "faceup": undefined,
    "id": "-graphic-test-id-5",
    "name": "Two of Clubs",
    "suit": "clubs",
    "x": -1,
    "y": -1,
  },
  {
    "cardid": "Lake",
    "faceup": true,
    "id": "Gap",
    "x": 0,
    "y": -1,
  },
  {
    "blank": false,
    "cardid": "-card-test-id-31",
    "face": "eight",
    "faceup": undefined,
    "id": "-graphic-test-id-31",
    "name": "Eight of Spades",
    "suit": "spades",
    "x": 1,
    "y": -1,
  },
  {
    "cardid": "Lake",
    "faceup": true,
    "id": "Gap",
    "x": -2,
    "y": 0,
  },
  {
    "blank": false,
    "cardid": "-card-test-id-44",
    "face": "queen",
    "faceup": undefined,
    "id": "-graphic-test-id-44",
    "name": "Queen of Hearts",
    "suit": "hearts",
    "x": -1,
    "y": 0,
  },
  {
    "cardid": "Lake",
    "faceup": true,
    "id": "Gap",
    "x": 0,
    "y": 1,
  },
  {
    "blank": false,
    "cardid": "-card-test-id-18",
    "face": "five",
    "faceup": undefined,
    "id": "-graphic-test-id-18",
    "name": "Five of Diamonds",
    "suit": "diamonds",
    "x": 1,
    "y": -2,
  },
  {
    "cardid": "Lake",
    "faceup": true,
    "id": "Gap",
    "x": -2,
    "y": -1,
  },
  {
    "blank": false,
    "cardid": "-card-test-id-4",
    "face": "two",
    "faceup": undefined,
    "id": "-graphic-test-id-4",
    "name": "Two of Hearts",
    "suit": "hearts",
    "x": -1,
    "y": 1,
  },
  {
    "cardid": "Lake",
    "faceup": true,
    "id": "Gap",
    "x": 2,
    "y": -2,
  },
  {
    "blank": false,
    "cardid": "-card-test-id-21",
    "face": "six",
    "faceup": undefined,
    "id": "-graphic-test-id-21",
    "name": "Six of Clubs",
    "suit": "clubs",
    "x": 0,
    "y": -2,
  },
  {
    "cardid": "Lake",
    "faceup": true,
    "id": "Gap",
    "x": -2,
    "y": -2,
  },
  {
    "blank": false,
    "cardid": "-card-test-id-50",
    "face": "king",
    "faceup": undefined,
    "id": "-graphic-test-id-50",
    "name": "King of Diamonds",
    "suit": "diamonds",
    "x": -1,
    "y": 2,
  },
  {
    "cardid": "Lake",
    "faceup": true,
    "id": "Gap",
    "x": 0,
    "y": -3,
  },
  {
    "blank": false,
    "cardid": "-card-test-id-7",
    "face": "two",
    "faceup": undefined,
    "id": "-graphic-test-id-7",
    "name": "Two of Spades",
    "suit": "spades",
    "x": 2,
    "y": -3,
  },
  {
    "cardid": "Lake",
    "faceup": true,
    "id": "Gap",
    "x": 1,
    "y": -3,
  },
  {
    "blank": false,
    "cardid": "-card-test-id-26",
    "face": "seven",
    "faceup": undefined,
    "id": "-graphic-test-id-26",
    "name": "Seven of Diamonds",
    "suit": "diamonds",
    "x": -2,
    "y": 2,
  },
  {
    "cardid": "Lake",
    "faceup": true,
    "id": "Gap",
    "x": -1,
    "y": -3,
  },
  {
    "blank": false,
    "cardid": "-card-test-id-11",
    "face": "three",
    "faceup": undefined,
    "id": "-graphic-test-id-11",
    "name": "Three of Spades",
    "suit": "spades",
    "x": 2,
    "y": -1,
  },
  {
    "cardid": "Lake",
    "faceup": true,
    "id": "Gap",
    "x": -1,
    "y": -2,
  },
  {
    "blank": false,
    "cardid": "-card-test-id-32",
    "face": "nine",
    "faceup": undefined,
    "id": "-graphic-test-id-32",
    "name": "Nine of Hearts",
    "suit": "hearts",
    "x": -2,
    "y": 3,
  },
  {
    "cardid": "Lake",
    "faceup": true,
    "id": "Gap",
    "x": 3,
    "y": -1,
  },
  {
    "blank": false,
    "cardid": "-card-test-id-39",
    "face": "ten",
    "faceup": undefined,
    "id": "-graphic-test-id-39",
    "name": "Ten of Spades",
    "suit": "spades",
    "x": 0,
    "y": 2,
  },
  {
    "cardid": "Lake",
    "faceup": true,
    "id": "Gap",
    "x": 2,
    "y": -4,
  },
  {
    "blank": false,
    "cardid": "-card-test-id-19",
    "face": "five",
    "faceup": undefined,
    "id": "-graphic-test-id-19",
    "name": "Five of Spades",
    "suit": "spades",
    "x": -3,
    "y": 1,
  },
  {
    "cardid": "Lake",
    "faceup": true,
    "id": "Gap",
    "x": -3,
    "y": 2,
  },
  {
    "blank": false,
    "cardid": "-card-test-id-16",
    "face": "five",
    "faceup": undefined,
    "id": "-graphic-test-id-16",
    "name": "Five of Hearts",
    "suit": "hearts",
    "x": -2,
    "y": 1,
  },
  {
    "cardid": "Lake",
    "faceup": true,
    "id": "Gap",
    "x": 3,
    "y": -4,
  },
  {
    "blank": false,
    "cardid": "-card-test-id-52",
    "face": "red",
    "faceup": undefined,
    "id": "-graphic-test-id-52",
    "name": "Red Joker",
    "suit": "joker",
    "x": -1,
    "y": 3,
  },
  {
    "cardid": "Lake",
    "faceup": true,
    "id": "Gap",
    "x": 1,
    "y": -4,
  },
  {
    "blank": false,
    "cardid": "-card-test-id-35",
    "face": "nine",
    "faceup": undefined,
    "id": "-graphic-test-id-35",
    "name": "Nine of Spades",
    "suit": "spades",
    "x": 3,
    "y": -2,
  },
  {
    "cardid": "Lake",
    "faceup": true,
    "id": "Gap",
    "x": 0,
    "y": 3,
  },
  {
    "blank": false,
    "cardid": "-card-test-id-40",
    "face": "jack",
    "faceup": undefined,
    "id": "-graphic-test-id-40",
    "name": "Jack of Hearts",
    "suit": "hearts",
    "x": -2,
    "y": 4,
  },
  {
    "cardid": "Lake",
    "faceup": true,
    "id": "Gap",
    "x": -4,
    "y": 2,
  },
  {
    "blank": false,
    "cardid": "-card-test-id-33",
    "face": "nine",
    "faceup": undefined,
    "id": "-graphic-test-id-33",
    "name": "Nine of Clubs",
    "suit": "clubs",
    "x": -3,
    "y": 4,
  },
  {
    "cardid": "Lake",
    "faceup": true,
    "id": "Gap",
    "x": -4,
    "y": 0,
  },
  {
    "blank": false,
    "cardid": "-card-test-id-34",
    "face": "nine",
    "faceup": undefined,
    "id": "-graphic-test-id-34",
    "name": "Nine of Diamonds",
    "suit": "diamonds",
    "x": -3,
    "y": 0,
  },
  {
    "cardid": "Lake",
    "faceup": true,
    "id": "Gap",
    "x": 1,
    "y": 2,
  },
  {
    "blank": false,
    "cardid": "-card-test-id-8",
    "face": "three",
    "faceup": undefined,
    "id": "-graphic-test-id-8",
    "name": "Three of Hearts",
    "suit": "hearts",
    "x": -1,
    "y": 4,
  },
  {
    "cardid": "Lake",
    "faceup": true,
    "id": "Gap",
    "x": 1,
    "y": 1,
  },
  {
    "blank": false,
    "cardid": "-card-test-id-6",
    "face": "two",
    "faceup": undefined,
    "id": "-graphic-test-id-6",
    "name": "Two of Diamonds",
    "suit": "diamonds",
    "x": -4,
    "y": 3,
  },
  {
    "cardid": "Lake",
    "faceup": true,
    "id": "Gap",
    "x": -3,
    "y": 5,
  },
  {
    "blank": false,
    "cardid": "-card-test-id-36",
    "face": "ten",
    "faceup": undefined,
    "id": "-graphic-test-id-36",
    "name": "Ten of Hearts",
    "suit": "hearts",
    "x": -3,
    "y": 3,
  },
  {
    "cardid": "Lake",
    "faceup": true,
    "id": "Gap",
    "x": -4,
    "y": 4,
  },
  {
    "blank": false,
    "cardid": "-card-test-id-23",
    "face": "six",
    "faceup": undefined,
    "id": "-graphic-test-id-23",
    "name": "Six of Spades",
    "suit": "spades",
    "x": 2,
    "y": 0,
  },
  {
    "cardid": "Lake",
    "faceup": true,
    "id": "Gap",
    "x": 3,
    "y": -3,
  },
  {
    "blank": false,
    "cardid": "-card-test-id-37",
    "face": "ten",
    "faceup": undefined,
    "id": "-graphic-test-id-37",
    "name": "Ten of Clubs",
    "suit": "clubs",
    "x": 4,
    "y": -2,
  },
  {
    "cardid": "Lake",
    "faceup": true,
    "id": "Gap",
    "x": 4,
    "y": -3,
  },
  {
    "blank": false,
    "cardid": "-card-test-id-29",
    "face": "eight",
    "faceup": undefined,
    "id": "-graphic-test-id-29",
    "name": "Eight of Clubs",
    "suit": "clubs",
    "x": -2,
    "y": 5,
  },
  {
    "cardid": "Lake",
    "faceup": true,
    "id": "Gap",
    "x": -4,
    "y": 5,
  },
  {
    "blank": false,
    "cardid": "-card-test-id-51",
    "face": "king",
    "faceup": undefined,
    "id": "-graphic-test-id-51",
    "name": "King of Spades",
    "suit": "spades",
    "x": -4,
    "y": -1,
  },
  {
    "cardid": "Lake",
    "faceup": true,
    "id": "Gap",
    "x": -4,
    "y": 1,
  },
  {
    "blank": false,
    "cardid": "-card-test-id-10",
    "face": "three",
    "faceup": undefined,
    "id": "-graphic-test-id-10",
    "name": "Three of Diamonds",
    "suit": "diamonds",
    "x": -1,
    "y": 5,
  },
  {
    "cardid": "Lake",
    "faceup": true,
    "id": "Gap",
    "x": 4,
    "y": -1,
  },
  {
    "blank": false,
    "cardid": "-card-test-id-45",
    "face": "queen",
    "faceup": undefined,
    "id": "-graphic-test-id-45",
    "name": "Queen of Clubs",
    "suit": "clubs",
    "x": -5,
    "y": 2,
  },
  {
    "cardid": "Lake",
    "faceup": true,
    "id": "Gap",
    "x": 2,
    "y": 1,
  },
  {
    "blank": false,
    "cardid": "-card-test-id-24",
    "face": "seven",
    "faceup": undefined,
    "id": "-graphic-test-id-24",
    "name": "Seven of Hearts",
    "suit": "hearts",
    "x": 5,
    "y": -3,
  },
  {
    "cardid": "Lake",
    "faceup": true,
    "id": "Gap",
    "x": -2,
    "y": 6,
  },
  {
    "blank": false,
    "cardid": "-card-test-id-15",
    "face": "four",
    "faceup": undefined,
    "id": "-graphic-test-id-15",
    "name": "Four of Spades",
    "suit": "spades",
    "x": -5,
    "y": -2,
  },
  {
    "cardid": "Lake",
    "faceup": true,
    "id": "Gap",
    "x": -6,
    "y": 3,
  },
  {
    "blank": false,
    "cardid": "-card-test-id-3",
    "face": "ace",
    "faceup": undefined,
    "id": "-graphic-test-id-3",
    "name": "Ace of Spades",
    "suit": "spades",
    "x": -5,
    "y": 3,
  },
  {
    "cardid": "Lake",
    "faceup": true,
    "id": "Gap",
    "x": -5,
    "y": -1,
  },
  {
    "blank": false,
    "cardid": "-card-test-id-46",
    "face": "queen",
    "faceup": undefined,
    "id": "-graphic-test-id-46",
    "name": "Queen of Diamonds",
    "suit": "diamonds",
    "x": -6,
    "y": -3,
  },
  {
    "cardid": "Lake",
    "faceup": true,
    "id": "Gap",
    "x": 6,
    "y": -3,
  },
  {
    "blank": false,
    "cardid": "-card-test-id-2",
    "face": "ace",
    "faceup": undefined,
    "id": "-graphic-test-id-2",
    "name": "Ace of Diamonds",
    "suit": "diamonds",
    "x": -6,
    "y": 2,
  },
  {
    "cardid": "Lake",
    "faceup": true,
    "id": "Gap",
    "x": -3,
    "y": -2,
  },
  {
    "blank": false,
    "cardid": "-card-test-id-9",
    "face": "three",
    "faceup": undefined,
    "id": "-graphic-test-id-9",
    "name": "Three of Clubs",
    "suit": "clubs",
    "x": -5,
    "y": 1,
  },
  {
    "cardid": "Lake",
    "faceup": true,
    "id": "Gap",
    "x": -5,
    "y": -3,
  },
  {
    "blank": false,
    "cardid": "-card-test-id-20",
    "face": "six",
    "faceup": undefined,
    "id": "-graphic-test-id-20",
    "name": "Six of Hearts",
    "suit": "hearts",
    "x": -1,
    "y": 6,
  },
  {
    "cardid": "Lake",
    "faceup": true,
    "id": "Gap",
    "x": -3,
    "y": -1,
  },
  {
    "blank": false,
    "cardid": "-card-test-id-49",
    "face": "king",
    "faceup": undefined,
    "id": "-graphic-test-id-49",
    "name": "King of Clubs",
    "suit": "clubs",
    "x": -6,
    "y": 0,
  },
  {
    "cardid": "Lake",
    "faceup": true,
    "id": "Gap",
    "x": 0,
    "y": 4,
  },
  {
    "blank": false,
    "cardid": "-card-test-id-42",
    "face": "jack",
    "faceup": undefined,
    "id": "-graphic-test-id-42",
    "name": "Jack of Diamonds",
    "suit": "diamonds",
    "x": -5,
    "y": 0,
  },
  {
    "cardid": "Lake",
    "faceup": true,
    "id": "Gap",
    "x": 5,
    "y": -4,
  },
  {
    "blank": false,
    "cardid": "-card-test-id-25",
    "face": "seven",
    "faceup": undefined,
    "id": "-graphic-test-id-25",
    "name": "Seven of Clubs",
    "suit": "clubs",
    "x": 0,
    "y": 6,
  },
  {
    "cardid": "Lake",
    "faceup": true,
    "id": "Gap",
    "x": -6,
    "y": 4,
  },
  {
    "blank": false,
    "cardid": "-card-test-id-53",
    "face": "black",
    "faceup": undefined,
    "id": "-graphic-test-id-53",
    "name": "Black Joker",
    "suit": "joker",
    "x": -5,
    "y": 4,
  },
  {
    "cardid": "Lake",
    "faceup": true,
    "id": "Gap",
    "x": 0,
    "y": 5,
  },
  {
    "blank": false,
    "cardid": "-card-test-id-28",
    "face": "eight",
    "faceup": undefined,
    "id": "-graphic-test-id-28",
    "name": "Eight of Hearts",
    "suit": "hearts",
    "x": -3,
    "y": 6,
  },
  {
    "cardid": "Lake",
    "faceup": true,
    "id": "Gap",
    "x": 3,
    "y": 0,
  },
  {
    "blank": false,
    "cardid": "-card-test-id-22",
    "face": "six",
    "faceup": undefined,
    "id": "-graphic-test-id-22",
    "name": "Six of Diamonds",
    "suit": "diamonds",
    "x": -6,
    "y": -1,
  },
  {
    "cardid": "Lake",
    "faceup": true,
    "id": "Gap",
    "x": -6,
    "y": 1,
  },
  {
    "blank": false,
    "cardid": "-card-test-id-14",
    "face": "four",
    "faceup": undefined,
    "id": "-graphic-test-id-14",
    "name": "Four of Diamonds",
    "suit": "diamonds",
    "x": 5,
    "y": -2,
  },
  {
    "cardid": "Lake",
    "faceup": true,
    "id": "Gap",
    "x": -2,
    "y": 7,
  },
  {
    "blank": false,
    "cardid": "-card-test-id-13",
    "face": "four",
    "faceup": undefined,
    "id": "-graphic-test-id-13",
    "name": "Four of Clubs",
    "suit": "clubs",
    "x": -6,
    "y": -2,
  },
  {
    "cardid": "Lake",
    "faceup": true,
    "id": "Gap",
    "x": -7,
    "y": 1,
  },
  {
    "blank": false,
    "cardid": "-card-test-id-1",
    "face": "ace",
    "faceup": undefined,
    "id": "-graphic-test-id-1",
    "name": "Ace of Clubs",
    "suit": "clubs",
    "x": -6,
    "y": 5,
  },
  {
    "cardid": "Lake",
    "faceup": true,
    "id": "Gap",
    "x": -7,
    "y": -2,
  },
  {
    "blank": false,
    "cardid": "-card-test-id-38",
    "face": "ten",
    "faceup": undefined,
    "id": "-graphic-test-id-38",
    "name": "Ten of Diamonds",
    "suit": "diamonds",
    "x": -7,
    "y": -1,
  },
  {
    "cardid": "Lake",
    "faceup": true,
    "id": "Gap",
    "x": -4,
    "y": 6,
  },
  {
    "blank": false,
    "cardid": "-card-test-id-48",
    "face": "king",
    "faceup": undefined,
    "id": "-graphic-test-id-48",
    "name": "King of Hearts",
    "suit": "hearts",
    "x": 1,
    "y": 6,
  },
  {
    "cardid": "Lake",
    "faceup": true,
    "id": "Gap",
    "x": -8,
    "y": 0,
  },
  {
    "blank": false,
    "cardid": "-card-test-id-47",
    "face": "queen",
    "faceup": undefined,
    "id": "-graphic-test-id-47",
    "name": "Queen of Spades",
    "suit": "spades",
    "x": 0,
    "y": 7,
  },
  {
    "cardid": "Lake",
    "faceup": true,
    "id": "Gap",
    "x": -3,
    "y": 7,
  },
  {
    "blank": false,
    "cardid": "-card-test-id-17",
    "face": "five",
    "faceup": undefined,
    "id": "-graphic-test-id-17",
    "name": "Five of Clubs",
    "suit": "clubs",
    "x": 4,
    "y": -4,
  },
  {
    "cardid": "Lake",
    "faceup": true,
    "id": "Gap",
    "x": 1,
    "y": 7,
  },
  {
    "blank": false,
    "cardid": "-card-test-id-27",
    "face": "seven",
    "faceup": undefined,
    "id": "-graphic-test-id-27",
    "name": "Seven of Spades",
    "suit": "spades",
    "x": -7,
    "y": -3,
  },
  {
    "cardid": "Lake",
    "faceup": true,
    "id": "Gap",
    "x": -7,
    "y": 3,
  },
  {
    "blank": false,
    "cardid": "-card-test-id-41",
    "face": "jack",
    "faceup": undefined,
    "id": "-graphic-test-id-41",
    "name": "Jack of Clubs",
    "suit": "clubs",
    "x": -5,
    "y": 5,
  },
  {
    "cardid": "Lake",
    "faceup": true,
    "id": "Gap",
    "x": 3,
    "y": -5,
  },
  {
    "blank": false,
    "cardid": "-card-test-id-0",
    "face": "ace",
    "faceup": undefined,
    "id": "-graphic-test-id-0",
    "name": "Ace of Hearts",
    "suit": "hearts",
    "x": 5,
    "y": -5,
  },
  {
    "cardid": "Lake",
    "faceup": true,
    "id": "Gap",
    "x": -7,
    "y": 6,
  },
  {
    "blank": false,
    "cardid": "-card-test-id-43",
    "face": "jack",
    "faceup": undefined,
    "id": "-graphic-test-id-43",
    "name": "Jack of Spades",
    "suit": "spades",
    "x": -6,
    "y": 6,
  },
  {
    "cardid": "Lake",
    "faceup": true,
    "id": "Gap",
    "x": 5,
    "y": -6,
  },
  {
    "blank": false,
    "cardid": "-card-test-id-30",
    "face": "eight",
    "faceup": undefined,
    "id": "-graphic-test-id-30",
    "name": "Eight of Diamonds",
    "suit": "diamonds",
    "x": -5,
    "y": 6,
  },
]
`;
